cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -760,13 +760,14 @@ private predicate variableWriteCand(IRBlock bb, int i, SourceVariable v) {
 }
 
 private predicate sourceVariableIsGlobal(
-  SourceVariable sv, Cpp::GlobalOrNamespaceVariable global, IRFunction func, int indirectionIndex
+  SourceVariable sv, GlobalLikeVariable global, IRFunction func, int indirectionIndex
 ) {
   exists(IRVariable irVar, BaseIRVariable base |
     sourceVariableHasBaseAndIndex(sv, base, indirectionIndex) and
     irVar = base.getIRVariable() and
     irVar.getEnclosingIRFunction() = func and
-    global = irVar.getAst()
+    global = irVar.getAst() and
+    not irVar instanceof IRDynamicInitializationFlag
```

Comment:
Should this be here, or should this actually be part of the `GlobalLikeVariable` definition? That seems to be more consistent with what we do in the IR library, or not?

Reply:
Good point! We can do that 👍

Reply:
Shouldn't it be possible to do something like:
```ql
class GlobalLikeVariable extends Variable {
  GlobalLikeVariable() {
    this instanceof Cpp::GlobalOrNamespaceVariable or
    this.(Cpp::StaticLocalVariable).hasDynamicInitialization()
  }
}
```
?

Reply:
Ah, I see what you mean. Hmmm.... We still need to treat dynamically initialized static variables as global variables though, right? Otherwise, we don't get flow in something like:
```cpp
void foo() {
  static int x = get_initial_value(); // dynamic initialization
  sink(x);
  x = source();
}
```

Reply:
But I thought we were already handling dynamically initialised static locals before this change?

Reply:
Not the "the value wraps around" part of static variables. On `main` we currently support the flow from the dynamic initialization to subsequent uses, but not the fact that the value is kept across calls.

Reply:
Then I don't understand the modelling of dynamically initialised static variables. How do we ensure that in data flow we consider functions with such variables in two ways? With and without the dynamic initialisation being executed?

Reply:
Also I don't understand why we need `not irVar instanceof IRDynamicInitializationFlag` then.

Reply:
For dynamically initialized static variables such as:
```cpp
void use(...);

void dynamic_init(int n) {
  static int s = n;
  use(s);
  s = 0;
}
```
we create this IR:
```
3| void dynamic_init(int)
3|   Block 0
3|     v3_1(void)        = EnterFunction           :
3|     m3_2(unknown)     = AliasedDefinition       :
3|     m3_3(unknown)     = InitializeNonLocal      :
3|     m3_4(unknown)     = Chi                     : total:m3_2, partial:m3_3
3|     r3_5(glval<int>)  = VariableAddress[n]      :
3|     m3_6(int)         = InitializeParameter[n]  : &:r3_5
4|     r4_1(glval<bool>) = VariableAddress[s#init] :
4|     r4_2(bool)        = Load[s#init]            : &:r4_1, ~m3_3
4|     v4_3(void)        = ConditionalBranch       : r4_2
-|   False -> Block 1
-|   True -> Block 2

4|   Block 1
4|     r4_4(glval<int>) = VariableAddress[s] :
4|     r4_5(glval<int>) = VariableAddress[n] :
4|     r4_6(int)        = Load[n]            : &:r4_5, m3_6
4|     m4_7(int)        = Store[s]           : &:r4_4, r4_6
4|     m4_8(unknown)    = Chi                : total:m3_4, partial:m4_7
4|     r4_9(bool)       = Constant[1]        :
4|     m4_10(bool)      = Store[s#init]      : &:r4_1, r4_9
4|     m4_11(unknown)   = Chi                : total:m4_8, partial:m4_10
-|   Goto -> Block 2

5|   Block 2
5|     m5_1(int)            = Phi                  : from 0:~m3_3, from 1:m4_7
5|     m5_2(unknown)        = Phi                  : from 0:~m3_4, from 1:~m4_11
5|     r5_3(glval<unknown>) = FunctionAddress[use] :
5|     r5_4(glval<int>)     = VariableAddress[s]   :
5|     r5_5(int)            = Load[s]              : &:r5_4, m5_1
5|     v5_6(void)           = Call[use]            : func:r5_3, 0:r5_5
5|     m5_7(unknown)        = ^CallSideEffect      : ~m5_2
5|     m5_8(unknown)        = Chi                  : total:m5_2, partial:m5_7
6|     r6_1(int)            = Constant[0]          :
6|     r6_2(glval<int>)     = VariableAddress[n]   :
6|     m6_3(int)            = Store[n]             : &:r6_2, r6_1
7|     v7_1(void)           = NoOp                 :
3|     v3_7(void)           = ReturnVoid           :
3|     v3_8(void)           = AliasedUse           : ~m5_8
3|     v3_9(void)           = ExitFunction         :
```
i.e., we branch on a synthetic `s#init` IRVariable and initialize `s` conditionally. So now that this PR also considers all static variables as a kind of global variable, we correctly model the fact that the initialization of `s` "wraps" around, and because the IR generates this "conditional" assignment, we get that the value at the point of calling `use` is either `n` or 0.

Reply:
So if we restrict `GlobalLikeVariable` with your suggestion we fail to wrap the final assignment `s = 0` around to the beginning of the function

Reply:
Just for completeness after syncing in person. The check on `IRDynamicInitializationFlag` ensures we do not picking up on IR variables like `s#init`.

Reply:
So, I'm happy with this.